mongofiles/mongofiles.go

```diff
@@ -57,185 +62,282 @@ type MongoFiles struct {
-		}
-	case Search, Put, Get, Delete:
-		if len(args) > 2 {
-			return fmt.Errorf("too many positional arguments")
+// Write data to GridFS Upload Stream. If this file has not been written to before, this function will open a new stream that must be closed.
+// Note: the go driver buffers data until it hits a chunk size before writing to the database, so if the amount of data is written < chunkSize,
+// the actual write to the database will occur in Close.
+func (file *gfsFile) Write(p []byte) (int, error) {
+	if file.upStream == nil {
+		rawBSON, err := bson.Marshal(file.Metadata)
+		if err != nil {
+			return 0, fmt.Errorf("could not marshal metadata to BSON: %v", err)
 		}
-		// also make sure the supporting argument isn't literally an
-		// empty string for example, mongofiles get ""
-		if len(args) == 1 || args[1] == "" {
-			return fmt.Errorf("'%v' argument missing", args[0])
+
+		doc, err := bsonx.ReadDoc(rawBSON)
+		if err != nil {
+			return 0, fmt.Errorf("could not read metadata to document: %v", err)
 		}
-		mf.FileName = args[1]
-	case GetID, DeleteID:
-		if len(args) > 2 {
-			return fmt.Errorf("too many positional arguments")
+
+		// TODO: remove this (GO-815)
+		objectId, ok := file.Id.(primitive.ObjectID)
+		if !ok {
+			return 0, fmt.Errorf("need to use objectid for _id")
 		}
-		if len(args) == 1 || args[1] == "" {
-			return fmt.Errorf("'%v' argument missing", args[0])
+
+		stream, err := file.mf.bucket.OpenUploadStreamWithID(objectId, file.Name, &driverOptions.UploadOptions{Metadata: doc})
+		if err != nil {
+			return 0, err
 		}
-		mf.Id = args[1]
-	case PutID:
-		if len(args) > 3 {
-			return fmt.Errorf("too many positional arguments")
+		file.upStream = stream
+	}
+
+	return file.upStream.Write(p)
+}
+
+// Reads data from GridFS download stream. If this file has not been read from before, this function will open a new stream that must be closed.
+func (file *gfsFile) Read(buf []byte) (int, error) {
+	if file.downStream == nil {
+		// TODO: remove this (GO-815)
+		objectId, ok := file.Id.(primitive.ObjectID)
+		if !ok {
+			return 0, fmt.Errorf("need to use objectid for _id")
 		}
-		if len(args) < 3 || args[1] == "" || args[2] == "" {
-			return fmt.Errorf("'%v' argument(s) missing", args[0])
+
+		stream, err := file.mf.bucket.OpenDownloadStream(objectId)
+		if err != nil {
+			return 0, fmt.Errorf("could not open download stream: %v", err)
 		}
-		mf.FileName = args[1]
-		mf.Id = args[2]
-	default:
-		return fmt.Errorf("'%v' is not a valid command", args[0])
+		file.downStream = stream
 	}
 
-	if mf.StorageOptions.GridFSPrefix == "" {
-		return fmt.Errorf("--prefix can not be blank")
+	return file.downStream.Read(buf)
+}
+
+// Deletes the corresponding GridFS file in the database and its chunks.
+// Note: this file must be closed if it had been written to before being deleted. Any download streams will be closed as part of this deletion.
+func (file *gfsFile) Delete() error {
+	if file.upStream != nil {
```

Comment:
this would only happen if we ourselves forget to close it, right?

Reply:
Yep.

Reply:
In such a case, it might be worth calling that out in a comment.